File: Power Query/PQ_Challenge_146.R
library(tidyverse)
library(readxl)

input = read_excel("Power Query/PQ_Challenge_146.xlsx", range = "A1:D14")
test  = read_excel("Power Query/PQ_Challenge_146.xlsx", range = "F1:I7")

result = input %>%
  group_by(Group) %>%
  mutate(Category = ifelse(Value == Threshold, "Equal",
                           ifelse(Value > Threshold, "High", "Low"))) %>%
  ungroup() %>%
  filter(Category != "Equal") %>%
  group_by(Group, Category) %>%
  mutate(valid = ifelse(Category == "High", min(Value), max(Value))) %>%
  ungroup() %>%
  filter(Value == valid) %>%
  select(-valid, -Category)

identical(result, test)
# [1] TRUE

File: Power Query/100-199/146/PQ_Challenge_146.py
import pandas as pd

input_data = pd.read_excel("PQ_Challenge_146.xlsx", usecols="A:D", nrows=14)
test = pd.read_excel("PQ_Challenge_146.xlsx", usecols="F:I", nrows=7)

result = input_data.copy()
result["Category"] = result.apply(
    lambda r: "Equal" if r["Value"] == r["Threshold"] else ("High" if r["Value"] > r["Threshold"] else "Low"),
    axis=1,
)
result = result[result["Category"] != "Equal"].copy()
result["valid"] = result.groupby(["Group", "Category"])["Value"].transform("min")
low_mask = result["Category"] == "Low"
result.loc[low_mask, "valid"] = result.loc[low_mask].groupby(["Group", "Category"])["Value"].transform("max")
result = result[result["Value"] == result["valid"]].drop(columns=["valid", "Category"]).reset_index(drop=True)

print(result.equals(test))
